// @material-ui/pickers/Picker/Picker.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var Calendar_1 = __importDefault(require("../DatePicker/components/Calendar"));
var YearSelection_1 = __importDefault(require("../DatePicker/components/YearSelection"));
var MonthSelection_1 = __importDefault(require("../DatePicker/components/MonthSelection"));
var useViews_1 = require("../_shared/hooks/useViews");
var useUtils_1 = require("../_shared/hooks/useUtils");
var styles_1 = require("@material-ui/core/styles");
var prop_types_1 = require("../constants/prop-types");
var TimePickerView_1 = require("../TimePicker/components/TimePickerView");
var viewsMap = {
    year: YearSelection_1.default,
    month: MonthSelection_1.default,
    date: Calendar_1.default,
    hours: TimePickerView_1.TimePickerView,
    minutes: TimePickerView_1.TimePickerView,
    seconds: TimePickerView_1.TimePickerView,
};
var useStyles = styles_1.makeStyles({
    pickerView: {
        minHeight: 305,
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'center',
    },
}, { name: 'MuiPickersBasePicker' });
exports.Picker = function (props) {
    var date = props.date, ampm = props.ampm, views = props.views, disableToolbar = props.disableToolbar, disablePast = props.disablePast, disableFuture = props.disableFuture, hideTabs = props.hideTabs, onChange = props.onChange, openTo = props.openTo, minutesStep = props.minutesStep, dateRangeIcon = props.dateRangeIcon, timeIcon = props.timeIcon, unparsedMinDate = props.minDate, unparsedMaxDate = props.maxDate, animateYearScrolling = props.animateYearScrolling, leftArrowIcon = props.leftArrowIcon, rightArrowIcon = props.rightArrowIcon, renderDay = props.renderDay, shouldDisableDate = props.shouldDisableDate, allowKeyboardControl = props.allowKeyboardControl, onMonthChange = props.onMonthChange, onYearChange = props.onYearChange, leftArrowButtonProps = props.leftArrowButtonProps, rightArrowButtonProps = props.rightArrowButtonProps, ToolbarComponent = props.ToolbarComponent, loadingIndicator = props.loadingIndicator;
    var utils = useUtils_1.useUtils();
    var classes = useStyles();
    var _a = useViews_1.useViews(views, openTo, onChange), openView = _a.openView, setOpenView = _a.setOpenView, handleChangeAndOpenNext = _a.handleChangeAndOpenNext;
    var minDate = React.useMemo(function () { return utils.date(unparsedMinDate); }, [unparsedMinDate, utils]);
    var maxDate = React.useMemo(function () { return utils.date(unparsedMaxDate); }, [unparsedMaxDate, utils]);
    return (React.createElement(React.Fragment, null,
        !disableToolbar && (React.createElement(ToolbarComponent, __assign({ date: date, onChange: onChange, setOpenView: setOpenView, openView: openView, hideTabs: hideTabs, dateRangeIcon: dateRangeIcon, timeIcon: timeIcon }, props))),
        React.createElement("div", { className: classes.pickerView },
            openView === 'year' && (React.createElement(YearSelection_1.default, { date: date, onChange: handleChangeAndOpenNext('month'), minDate: minDate, maxDate: maxDate, disablePast: disablePast, disableFuture: disableFuture, onYearChange: onYearChange, animateYearScrolling: animateYearScrolling })),
            openView === 'month' && (React.createElement(MonthSelection_1.default, { date: date, onChange: handleChangeAndOpenNext('date'), minDate: minDate, maxDate: maxDate, disablePast: disablePast, disableFuture: disableFuture, onMonthChange: onMonthChange })),
            openView === 'date' && (React.createElement(Calendar_1.default, { date: date, onChange: handleChangeAndOpenNext('hours'), onMonthChange: onMonthChange, disablePast: disablePast, disableFuture: disableFuture, minDate: minDate, maxDate: maxDate, leftArrowIcon: leftArrowIcon, leftArrowButtonProps: leftArrowButtonProps, rightArrowIcon: rightArrowIcon, rightArrowButtonProps: rightArrowButtonProps, renderDay: renderDay, shouldDisableDate: shouldDisableDate, allowKeyboardControl: allowKeyboardControl, loadingIndicator: loadingIndicator })),
            (openView === 'hours' || openView === 'minutes' || openView === 'seconds') && (React.createElement(TimePickerView_1.TimePickerView, { date: date, ampm: ampm, type: openView, minutesStep: minutesStep, onHourChange: handleChangeAndOpenNext('minutes'), onMinutesChange: handleChangeAndOpenNext('seconds'), onSecondsChange: handleChangeAndOpenNext(null) })))));
};
exports.Picker.defaultProps = __assign({}, prop_types_1.datePickerDefaultProps, { views: Object.keys(viewsMap) });
