// @material-ui/pickers/_shared/hooks/useKeyboardPickerState.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var useUtils_1 = require("./useUtils");
var react_1 = require("react");
var text_field_helper_1 = require("../../_helpers/text-field-helper");
var usePickerState_1 = require("./usePickerState");
function parseInputString(value, utils, format) {
    try {
        return utils.parse(value, format);
    }
    catch (_a) {
        return null;
    }
}
function useKeyboardPickerState(props, options) {
    var utils = useUtils_1.useUtils();
    var format = props.format || options.getDefaultFormat();
    var _a = react_1.useState(text_field_helper_1.getDisplayDate(props.value, format, utils, props.value === null, props)), innerInputValue = _a[0], setInnerInputValue = _a[1];
    var dateValue = props.inputValue
        ? parseInputString(props.inputValue, utils, format)
        : props.value;
    react_1.useEffect(function () {
        if (props.value === null || utils.isValid(props.value)) {
            setInnerInputValue(text_field_helper_1.getDisplayDate(props.value, format, utils, props.value === null, props));
        }
    }, [format, props, props.value, utils]);
    function handleChange(date) {
        var dateString = date === null ? null : utils.format(date, format);
        props.onChange(date, dateString);
    }
    var _b = usePickerState_1.usePickerState(__assign({}, props, { value: dateValue, onChange: handleChange }), options), innerInputProps = _b.inputProps, wrapperProps = _b.wrapperProps, pickerProps = _b.pickerProps;
    var inputProps = react_1.useMemo(function () { return (__assign({}, innerInputProps, { format: wrapperProps.format, inputValue: props.inputValue || innerInputValue, onChange: function (value) {
            setInnerInputValue(value);
            var date = value === '' ? null : utils.parse(value, wrapperProps.format);
            props.onChange(date, value);
        } })); }, [innerInputProps, innerInputValue, props, utils, wrapperProps.format]);
    return {
        inputProps: inputProps,
        wrapperProps: wrapperProps,
        pickerProps: pickerProps,
    };
}
exports.useKeyboardPickerState = useKeyboardPickerState;
